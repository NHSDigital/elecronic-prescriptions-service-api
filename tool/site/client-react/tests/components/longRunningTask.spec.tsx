import LongRunningTask from "../../src/components/longRunningTask"
import {render, screen} from "@testing-library/react"
import React from "react"
import pretty from "pretty"
import userEvent from "@testing-library/user-event"

test("Shows loading message while task runs", async () => {
  const mockTask = jest.fn()
  mockTask.mockReturnValue(new Promise(jest.fn()))

  const ui = <LongRunningTask<Record<string, string>> task={mockTask} loadingMessage="Some loading message">
    {result => <span>{result.data}</span>}
  </LongRunningTask>
  const {container} = render(ui)

  await screen.findByText("Some loading message")
  expect(mockTask).toHaveBeenCalledTimes(1)
  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test("Renders children if task resolves", async () => {
  const mockTask = jest.fn()
  mockTask.mockResolvedValue({data: "Some data from the backend"})

  const ui = <LongRunningTask<Record<string, string>> task={mockTask} loadingMessage="Loading">
    {result => <span>{result.data}</span>}
  </LongRunningTask>
  const {container} = render(ui)

  await screen.findByText("Some data from the backend")
  expect(mockTask).toHaveBeenCalledTimes(1)
  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test.each([
  ["an error with a message", new Error("Some error message"), "Some error message"],
  ["an error without a message", new Error(), "Unknown error"],
  ["a string", "Some error message", "Some error message"],
  ["anything else", null, "Unknown error"]
])("Displays an error if task rejects with %s", async (desc: string, rejectedValue: unknown, expectedText: string) => {
  const mockTask = jest.fn()
  mockTask.mockRejectedValue(rejectedValue)

  const ui = <LongRunningTask<Record<string, string>> task={mockTask} loadingMessage="Loading">
    {result => <span>{result.data}</span>}
  </LongRunningTask>
  const {container} = render(ui)

  await screen.findByText(expectedText)
  expect(mockTask).toHaveBeenCalledTimes(1)
  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test("Error page includes a back button with the provided href", async () => {
  const mockTask = jest.fn()
  mockTask.mockRejectedValue(new Error("Some error message"))

  const ui = (
    <LongRunningTask<Record<string, string>> task={mockTask} loadingMessage="Loading" back="http://example.com/">
      {result => <span>{result.data}</span>}
    </LongRunningTask>
  )
  const {container} = render(ui)

  const linkButton = await screen.findByText<HTMLAnchorElement>("Back")
  expect(linkButton.href).toEqual("http://example.com/")
  expect(mockTask).toHaveBeenCalledTimes(1)
  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test("Error page includes a back button with the provided onclick handler", async () => {
  const mockTask = jest.fn()
  mockTask.mockRejectedValue(new Error("Some error message"))
  const mockBack = jest.fn()

  const ui = (
    <LongRunningTask<Record<string, string>> task={mockTask} loadingMessage="Loading" back={mockBack}>
      {result => <span>{result.data}</span>}
    </LongRunningTask>
  )
  const {container} = render(ui)

  const button = await screen.findByText<HTMLButtonElement>("Back")
  userEvent.click(button)
  expect(mockTask).toHaveBeenCalledTimes(1)
  expect(mockBack).toHaveBeenCalledTimes(1)
  expect(pretty(container.innerHTML)).toMatchSnapshot()
})
