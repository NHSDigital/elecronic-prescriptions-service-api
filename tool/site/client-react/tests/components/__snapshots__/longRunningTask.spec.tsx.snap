// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Displays an error if task rejects with a string 1`] = `
"<h1 class=\\"nhsuk-label-wrapper\\"><label class=\\"nhsuk-label nhsuk-label--xl\\">Error</label></h1><span class=\\"nhsuk-error-message\\" role=\\"alert\\"><span class=\\"nhsuk-u-visually-hidden\\">Error: </span>Some error message</span>
<div class=\\"sc-bdvvtL jjZBkx\\"><a class=\\"nhsuk-button nhsuk-button--secondary\\" role=\\"button\\" aria-disabled=\\"false\\" draggable=\\"false\\" href=\\"/\\">Back</a></div>"
`;

exports[`Displays an error if task rejects with an error with a message 1`] = `
"<h1 class=\\"nhsuk-label-wrapper\\"><label class=\\"nhsuk-label nhsuk-label--xl\\">Error</label></h1><span class=\\"nhsuk-error-message\\" role=\\"alert\\"><span class=\\"nhsuk-u-visually-hidden\\">Error: </span>Some error message</span>
<div class=\\"sc-bdvvtL jjZBkx\\"><a class=\\"nhsuk-button nhsuk-button--secondary\\" role=\\"button\\" aria-disabled=\\"false\\" draggable=\\"false\\" href=\\"/\\">Back</a></div>"
`;

exports[`Displays an error if task rejects with an error without a message 1`] = `
"<h1 class=\\"nhsuk-label-wrapper\\"><label class=\\"nhsuk-label nhsuk-label--xl\\">Error</label></h1><span class=\\"nhsuk-error-message\\" role=\\"alert\\"><span class=\\"nhsuk-u-visually-hidden\\">Error: </span>Unknown error.</span>
<div class=\\"sc-bdvvtL jjZBkx\\"><a class=\\"nhsuk-button nhsuk-button--secondary\\" role=\\"button\\" aria-disabled=\\"false\\" draggable=\\"false\\" href=\\"/\\">Back</a></div>"
`;

exports[`Displays an error if task rejects with anything else 1`] = `
"<h1 class=\\"nhsuk-label-wrapper\\"><label class=\\"nhsuk-label nhsuk-label--xl\\">Error</label></h1><span class=\\"nhsuk-error-message\\" role=\\"alert\\"><span class=\\"nhsuk-u-visually-hidden\\">Error: </span>Unknown error.</span>
<div class=\\"sc-bdvvtL jjZBkx\\"><a class=\\"nhsuk-button nhsuk-button--secondary\\" role=\\"button\\" aria-disabled=\\"false\\" draggable=\\"false\\" href=\\"/\\">Back</a></div>"
`;

exports[`Error page includes a back button with the provided href 1`] = `
"<h1 class=\\"nhsuk-label-wrapper\\"><label class=\\"nhsuk-label nhsuk-label--xl\\">Error</label></h1><span class=\\"nhsuk-error-message\\" role=\\"alert\\"><span class=\\"nhsuk-u-visually-hidden\\">Error: </span>Some error message</span>
<div class=\\"sc-bdvvtL jjZBkx\\"><a class=\\"nhsuk-button nhsuk-button--secondary\\" role=\\"button\\" aria-disabled=\\"false\\" draggable=\\"false\\" href=\\"http://example.com/\\">Back</a></div>"
`;

exports[`Error page includes a back button with the provided onclick handler 1`] = `
"<h1 class=\\"nhsuk-label-wrapper\\"><label class=\\"nhsuk-label nhsuk-label--xl\\">Error</label></h1><span class=\\"nhsuk-error-message\\" role=\\"alert\\"><span class=\\"nhsuk-u-visually-hidden\\">Error: </span>Some error message</span>
<div class=\\"sc-bdvvtL jjZBkx\\"><button class=\\"nhsuk-button nhsuk-button--secondary\\" aria-disabled=\\"false\\" type=\\"submit\\">Back</button></div>"
`;

exports[`Renders children if task resolves 1`] = `"<span>Some data from the backend</span>"`;

exports[`Shows loading message while task runs 1`] = `"<h1 class=\\"nhsuk-label-wrapper\\"><label class=\\"nhsuk-label nhsuk-label--xl\\">Loading...</label></h1><label class=\\"nhsuk-label\\">Some loading message</label>"`;
